File: 백준/Silver/9012. 괄호/괄호.java
import java.io.*;
import java.util.Stack;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int T = Integer.parseInt(br.readLine());
        
        for(int i = 0; i < T; i++) {
            Stack<Character> s = new Stack<>();
            
            for(char c : br.readLine().toCharArray()) {
                if(c == '(') {
                    s.push(c);
                }
                else {
                    if(!s.empty()) {
                        s.pop();
                    }
                    else {
                        s.push(' ');
                        break;
                    }
                }
            }

            if(s.empty()) {
                bw.write("YES");
                bw.newLine();
            }
            else {
                bw.write("NO");
                bw.newLine();
            }
        }


        bw.flush();
        bw.close();
        br.close();
    }
}

File: 백준/Silver/9012. 괄호/괄호.kt
import java.util.*

fun main() {
    val T = readln().toInt()

    repeat(T) {
        val input = readln()
        val stack = Stack<Char>()
        for(c in input.toCharArray()) {
            when(c) {
                '(' -> stack.push(c)
                ')' -> if(!stack.empty()) {
                    stack.pop()
                }
                else {
                    stack.push(' ')
                    break
                }
            }
        }

        when {
            stack.empty() -> println("YES")
            else -> println("NO")
        }
    }
}
